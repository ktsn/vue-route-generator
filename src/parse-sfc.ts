import type { parseComponent as _parserV2 } from 'vue-template-compiler'
import type { parse as _parserV3 } from '@vue/compiler-sfc'

export interface ParseResult {
  customBlocks: CustomBlock[]
}

export interface CustomBlock {
  type: string
  content: string
}

export function parseSFC(code: string): ParseResult {
  try {
    const parserV2: typeof _parserV2 = require('vue-template-compiler')
      .parseComponent
    return parserV2(code, {
      pad: 'space',
    })
  } catch {
    try {
      const parserV3: typeof _parserV3 = require('@vue/compiler-sfc').parse
      return parserV3(code, {
        pad: 'space',
      }).descriptor
    } catch {
      throw new Error(
        '[vue-route-generator] Either "vue-template-compiler" or "@vue/compiler-sfc" is required.'
      )
    }
  }
}
