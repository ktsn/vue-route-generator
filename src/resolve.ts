import { NestedMap, setToMap } from './nested-map'
import { parseSFC } from './parse-sfc'

export interface PageMeta {
  name: string
  chunkName: string
  specifier: string
  path: string
  pathSegments: string[]
  component: string
  children?: PageMeta[]
  routeMeta?: any
  route?: any
}

interface FileError extends Error {
  file?: string
}

const routeMetaName = 'route-meta'
const routeBlockName = 'route'

export function resolveRoutePaths(
  paths: string[],
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string
): PageMeta[] {
  const map: NestedMap<string[]> = {}

  const splitted = paths.map((p) => p.split('/'))
  splitted.forEach((path) => {
    setToMap(map, pathToMapPath(path), path)
  })

  return pathMapToMeta(map, importPrefix, nested, readFile)
}

function pathMapToMeta(
  map: NestedMap<string[]>,
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string,
  parentDepth: number = 0
): PageMeta[] {
  if (map.value) {
    const path = map.value

    const meta: PageMeta = {
      name: pathToName(path),
      chunkName: pathToChunkName(path),
      specifier: pathToSpecifier(path),
      path: pathToRoute(path, parentDepth, nested),
      pathSegments: toActualPath(path),
      component: importPrefix + path.join('/'),
    }

    const content = readFile(path.join('/'))
    const parsed = parseSFC(content)
    const routeMetaBlock = parsed.customBlocks.find(
      (b) => b.type === routeMetaName
    )
    const routeBlock = parsed.customBlocks.find(
      (b) => b.type === routeBlockName
    )

    // Deprecated. Will be removed in a later version
    if (routeMetaBlock) {
      console.warn(
        '<route-meta> custom block is deprecated. Use <route> block instead. Found in ' +
          path.join('/')
      )
      meta.routeMeta = tryParseCustomBlock(
        routeMetaBlock.content,
        path,
        'route-meta'
      )
    }

    if (routeBlock) {
      meta.route = tryParseCustomBlock(routeBlock.content, path, 'route')
    }

    if (map.children) {
      meta.children = pathMapChildrenToMeta(
        map.children,
        importPrefix,
        nested,
        readFile,
        meta.pathSegments.length
      )
    }

    return [meta]
  }

  return map.children
    ? pathMapChildrenToMeta(
        map.children,
        importPrefix,
        nested,
        readFile,
        parentDepth
      )
    : []
}

function routePathComparator(a: string[], b: string[]): number {
  const a0 = a[0]
  const b0 = b[0]

  if (!a0 || !b0) {
    return a.length - b.length
  }

  const aOrder = isDynamicRoute(a0) ? 1 : 0
  const bOrder = isDynamicRoute(b0) ? 1 : 0
  const order = aOrder - bOrder

  return order !== 0 ? order : routePathComparator(a.slice(1), b.slice(1))
}

function pathMapChildrenToMeta(
  children: Map<string, NestedMap<string[]>>,
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string,
  parentDepth: number
): PageMeta[] {
  return Array.from(children.values())
    .reduce<PageMeta[]>((acc, value) => {
      return acc.concat(
        pathMapToMeta(value, importPrefix, nested, readFile, parentDepth)
      )
    }, [])
    .sort((a, b) => {
      // Prioritize static routes than dynamic routes
      return routePathComparator(a.pathSegments, b.pathSegments)
    })
}

function tryParseCustomBlock(
  content: string,
  filePath: string[],
  blockName: string
): any {
  try {
    return JSON.parse(content)
  } catch (err) {
    const joinedPath = filePath.join('/')
    const wrapped: FileError = new Error(
      `Invalid json format of <${blockName}> content in ${joinedPath}\n` +
        err.message
    )

    // Store file path to provide useful information to downstream tools
    // like friendly-errors-webpack-plugin
    wrapped.file = joinedPath

    throw wrapped
  }
}

function isDynamicRoute(segment: string): boolean {
  return segment[0] === ':'
}

function isOmittable(segment: string): boolean {
  return segment === 'index'
}

/**
 * - Remove `.vue` from the last path
 * - Omit if the last segument is `index`
 * - Convert dynamic route to `:param` format
 */
function toActualPath(segments: string[]): string[] {
  const lastIndex = segments.length - 1
  const last = basename(segments[lastIndex])
  segments = segments.slice(0, -1).concat(last)

  return segments
    .map((s, i) => {
      if (s[0] === '_') {
        const suffix = lastIndex === i ? '?' : ''
        return ':' + s.slice(1) + suffix
      } else {
        return s
      }
    })
    .filter((s) => !isOmittable(s))
}

function pathToMapPath(segments: string[]): string[] {
  const last = segments[segments.length - 1]
  return segments.slice(0, -1).concat(basename(last))
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments
    .slice(0, -1)
    .concat(basename(last))
    .filter((s) => !isOmittable(s))

  if (segments.length === 0) {
    return 'index'
  }

  return segments
    .map((s) => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToChunkName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments.slice(0, -1).concat(basename(last))

  return segments
    .map((s) => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const last = segments[segments.length - 1]
  const replaced = segments
    .slice(0, -1)
    .concat(basename(last))
    .join('_')
    .replace(/[^a-zA-Z0-9]/g, '_')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(
  segments: string[],
  parentDepth: number,
  nested: boolean
): string {
  const prefix = nested || parentDepth > 0 ? '' : '/'
  return prefix + toActualPath(segments).slice(parentDepth).join('/')
}

function basename(filename: string): string {
  return filename.replace(/\.[^.]+$/g, '')
}
