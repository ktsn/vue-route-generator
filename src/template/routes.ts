import { transformSync } from '@babel/core'
import { PageMeta } from '../resolve'

function isAllowedRouteOption(key: string): boolean {
  return !['name', 'meta', 'path', 'component'].includes(key)
}

function createChildrenRoute(
  children: PageMeta[],
  inlineRouteBlock: boolean
): string {
  return `,children: [${children
    .map((c) => createRoute(c, inlineRouteBlock))
    .join(',')}]`
}

function createRoute(meta: PageMeta, inlineRouteBlock: boolean): string {
  const children = !meta.children
    ? ''
    : createChildrenRoute(meta.children, inlineRouteBlock)
  const route = meta.route

  // If default child is exists, the route should not have a name.
  const routeName =
    meta.children && meta.children.some((m) => m.path === '')
      ? ''
      : `name: '${route?.content?.name ?? meta.name}',`

  const routeMeta = meta.routeMeta
    ? ',meta: ' + JSON.stringify(meta.routeMeta, null, 2)
    : route?.content?.meta
    ? ',meta: ' + JSON.stringify(route.content.meta, null, 2)
    : ''

  const otherOptions = Object.keys(route?.content ?? {})
    .filter(isAllowedRouteOption)
    .map((key) => `,${key}: ${JSON.stringify(route?.content[key])}`)
    .join(',')

  const routeBlockSpread =
    !inlineRouteBlock && meta.route ? `...${meta.specifier}___route,` : ''

  return `
  {
    ${routeName}${routeBlockSpread}
    path: '${meta.path}',
    component: ${meta.specifier}${routeMeta}${otherOptions}${children}
  }`
}

function createImport(
  meta: PageMeta,
  dynamic: boolean,
  chunkNamePrefix: string,
  inlineRouteBlock: boolean
): string {
  let code = dynamic
    ? `function ${meta.specifier}() { return import(/* webpackChunkName: "${chunkNamePrefix}${meta.chunkName}" */ '${meta.component}') }`
    : `import ${meta.specifier} from '${meta.component}'`

  if (meta.route && !inlineRouteBlock) {
    code += `\nimport ${meta.specifier}___route from '${meta.component}?vue&vueRouteGenerator&type=custom&blockType=route&index=${meta.route.index}&lang=${meta.route.lang}'`
  }

  return meta.children
    ? [code]
        .concat(
          meta.children.map((child) =>
            createImport(child, dynamic, chunkNamePrefix, inlineRouteBlock)
          )
        )
        .join('\n')
    : code
}

export function createRoutes(
  meta: PageMeta[],
  dynamic: boolean,
  chunkNamePrefix: string,
  inlineRouteBlock: boolean
): string {
  const imports = meta
    .map((m) => createImport(m, dynamic, chunkNamePrefix, inlineRouteBlock))
    .join('\n')
  const routes = meta.map((m) => createRoute(m, inlineRouteBlock)).join(',')
  const code = `${imports}\n\nexport default [${routes}]`

  const result = transformSync(code, {
    presets: [['@babel/preset-env', { modules: false }]],
  })

  if (!result?.code) {
    throw new Error('Error during transform')
  }

  return result.code
}
