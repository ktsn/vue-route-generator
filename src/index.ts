import * as fs from 'fs'
import * as path from 'path'
import * as fg from 'fast-glob'
import { createRoutes } from './template/routes'
import { resolveRoutePaths } from './resolve'

export interface GenerateConfig {
  pages: string
  importPrefix?: string
  dynamicImport?: boolean
  chunkNamePrefix?: string
  nested?: boolean
  inlineRouteBlock?: boolean
}

export function generateRoutes({
  pages,
  importPrefix = '@/pages/',
  dynamicImport = true,
  chunkNamePrefix = '',
  nested = false,
  inlineRouteBlock = true,
}: GenerateConfig): string {
  const patterns = ['**/*.vue', '!**/__*__.vue', '!**/__*__/**']

  const pagePaths = fg.sync(patterns, {
    cwd: pages,
    onlyFiles: true,
  })

  const metaList = resolveRoutePaths(pagePaths, {
    importPrefix,
    nested,
    inlineRouteBlock,
    readFile: (file) => {
      return fs.readFileSync(path.join(pages, file), 'utf8')
    },
  })

  return createRoutes(
    metaList,
    dynamicImport,
    chunkNamePrefix,
    inlineRouteBlock
  )
}
