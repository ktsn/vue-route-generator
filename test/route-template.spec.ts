import { createRoutes } from '../src/template/routes'
import { PageMeta } from '../src/resolve'

describe('Route template', () => {
  it('should generate routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
      },
      {
        name: 'bar',
        chunkName: 'bar',
        specifier: 'Bar',
        path: '/bar',
        pathSegments: ['bar'],
        component: '@/pages/bar.vue',
      },
    ]

    expect(createRoutes(meta, true, '', true)).toMatchSnapshot()
  })

  it('should generate nested routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
        children: [
          {
            name: 'bar',
            chunkName: 'bar',
            specifier: 'FooBar',
            path: 'bar',
            pathSegments: ['foo', 'bar'],
            component: '@/pages/bar.vue',
          },
          {
            name: 'baz',
            chunkName: 'baz',
            specifier: 'FooBaz',
            path: 'baz',
            pathSegments: ['foo', 'baz'],
            component: '@/pages/baz.vue',
          },
        ],
      },
    ]

    expect(createRoutes(meta, true, '', true)).toMatchSnapshot()
  })

  it('should generate static import code', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
      },
      {
        name: 'bar',
        chunkName: 'bar',
        specifier: 'Bar',
        path: '/bar',
        pathSegments: ['bar'],
        component: '@/pages/bar.vue',
      },
    ]

    expect(createRoutes(meta, false, '', true)).toMatchSnapshot()
  })

  it('should generate route meta', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
        routeMeta: {
          title: 'Hello',
        },
      },
    ]

    expect(createRoutes(meta, false, '', true)).toMatchSnapshot()
  })

  it('should merge route block into route record', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
        route: {
          lang: 'json',
          index: 0,
          content: {
            name: 'Test',
            meta: {
              title: 'Hello',
            },
            props: true,
          },
        },
      },
    ]

    expect(createRoutes(meta, false, '', true)).toMatchSnapshot()
  })

  it('should import and merge route block if inlineRouteBlock is false', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
        route: {
          lang: 'json',
          index: 0,
        },
      },
    ]

    expect(createRoutes(meta, false, '', false)).toMatchSnapshot()
  })

  it('should configure chunk name prefix', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
      },
      {
        name: 'bar',
        chunkName: 'bar',
        specifier: 'Bar',
        path: '/bar',
        pathSegments: ['bar'],
        component: '@/pages/bar.vue',
      },
    ]

    expect(createRoutes(meta, true, 'page-', true)).toMatchSnapshot()
  })

  it('should pick chunkName as a webpack chunk name', () => {
    const meta: PageMeta[] = [
      {
        name: 'users',
        chunkName: 'users-index',
        specifier: 'users_index',
        path: '/users',
        pathSegments: ['users'],
        component: '@/pages/users/index.vue',
      },
    ]

    expect(createRoutes(meta, true, 'page-', true)).toMatchSnapshot()
  })

  it('should not include name if the route has a default child', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        chunkName: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/pages/foo.vue',
        children: [
          {
            name: 'foo',
            chunkName: 'foo-index',
            specifier: 'FooIndex',
            path: '',
            pathSegments: ['foo'],
            component: '@/pages/foo/index.vue',
          },
        ],
      },
    ]

    expect(createRoutes(meta, true, '', true)).toMatchSnapshot()
  })
})
