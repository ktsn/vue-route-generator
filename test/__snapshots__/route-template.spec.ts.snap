// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route template should configure chunk name prefix 1`] = `
"function Foo() {
  return import(
  /* webpackChunkName: \\"page-foo\\" */
  '@/pages/foo.vue');
}

function Bar() {
  return import(
  /* webpackChunkName: \\"page-bar\\" */
  '@/pages/bar.vue');
}

export default [{
  name: 'foo',
  path: '/foo',
  component: Foo
}, {
  name: 'bar',
  path: '/bar',
  component: Bar
}];"
`;

exports[`Route template should generate nested routes 1`] = `
"function Foo() {
  return import(
  /* webpackChunkName: \\"foo\\" */
  '@/pages/foo.vue');
}

function FooBar() {
  return import(
  /* webpackChunkName: \\"bar\\" */
  '@/pages/bar.vue');
}

function FooBaz() {
  return import(
  /* webpackChunkName: \\"baz\\" */
  '@/pages/baz.vue');
}

export default [{
  name: 'foo',
  path: '/foo',
  component: Foo,
  children: [{
    name: 'bar',
    path: 'bar',
    component: FooBar
  }, {
    name: 'baz',
    path: 'baz',
    component: FooBaz
  }]
}];"
`;

exports[`Route template should generate route meta 1`] = `
"import Foo from '@/pages/foo.vue';
export default [{
  name: 'foo',
  path: '/foo',
  component: Foo,
  meta: {
    \\"title\\": \\"Hello\\"
  }
}];"
`;

exports[`Route template should generate routes 1`] = `
"function Foo() {
  return import(
  /* webpackChunkName: \\"foo\\" */
  '@/pages/foo.vue');
}

function Bar() {
  return import(
  /* webpackChunkName: \\"bar\\" */
  '@/pages/bar.vue');
}

export default [{
  name: 'foo',
  path: '/foo',
  component: Foo
}, {
  name: 'bar',
  path: '/bar',
  component: Bar
}];"
`;

exports[`Route template should generate static import code 1`] = `
"import Foo from '@/pages/foo.vue';
import Bar from '@/pages/bar.vue';
export default [{
  name: 'foo',
  path: '/foo',
  component: Foo
}, {
  name: 'bar',
  path: '/bar',
  component: Bar
}];"
`;

exports[`Route template should import and merge route block if inlineRouteBlock is false 1`] = `
"function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import Foo from '@/pages/foo.vue';
import Foo___route from '@/pages/foo.vue?vue&vueRouteGenerator&type=custom&blockType=route&index=0&lang=json';
export default [_objectSpread(_objectSpread({
  name: 'foo'
}, Foo___route), {}, {
  path: '/foo',
  component: Foo
})];"
`;

exports[`Route template should merge route block into route record 1`] = `
"import Foo from '@/pages/foo.vue';
export default [{
  name: 'Test',
  path: '/foo',
  component: Foo,
  meta: {
    \\"title\\": \\"Hello\\"
  },
  props: true
}];"
`;

exports[`Route template should not include name if the route has a default child 1`] = `
"function Foo() {
  return import(
  /* webpackChunkName: \\"foo\\" */
  '@/pages/foo.vue');
}

function FooIndex() {
  return import(
  /* webpackChunkName: \\"foo-index\\" */
  '@/pages/foo/index.vue');
}

export default [{
  path: '/foo',
  component: Foo,
  children: [{
    name: 'foo',
    path: '',
    component: FooIndex
  }]
}];"
`;

exports[`Route template should pick chunkName as a webpack chunk name 1`] = `
"function users_index() {
  return import(
  /* webpackChunkName: \\"page-users-index\\" */
  '@/pages/users/index.vue');
}

export default [{
  name: 'users',
  path: '/users',
  component: users_index
}];"
`;
